Inability to write annotations on the outer class portion of an inner class name

Dataflow seems to break with inner class types.
```java
import org.checkerframework.checker.tainting.qual.*;

public class C {
    void method(@Tainted Inner param) {
        @Tainted C. @Untainted Inner local = param;
    }

    class Inner {
    }
}
```
I get this error,
```
C.java:5: error: [type.invalid.conflicting.annos] invalid type: conflicting annotations [@Tainted, @Untainted] in type "@Tainted C.@Tainted @Untainted Inner"
        @Tainted C. @Untainted Inner local = param;
                                     ^
```
I would expect to instead get an invalid assignment error. It seems dataflow tries to replace the inner type but doesn't remove the old annotation.

--- a/checker/tests/tainting/Issue3562.java
+++ b/checker/tests/tainting/Issue3562.java
@@ -1,0 +1,8 @@
+import org.checkerframework.checker.tainting.qual.*;
+
+public class Issue3562 {
+    // This used to issue type.invalid.conflicting.annos
+    @Tainted Issue3562.@Untainted Inner field;
+
+    class Inner {}
+}

--- a/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
@@ -6,11 +6,15 @@ import com.sun.source.tree.MethodTree;
 import com.sun.source.tree.Tree;
 import com.sun.source.tree.Tree.Kind;
 import com.sun.source.tree.VariableTree;
+import java.util.ArrayList;
 import java.util.List;
 import javax.lang.model.element.AnnotationMirror;
 import javax.lang.model.element.Element;
 import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeKind;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedDeclaredType;
 import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
 import org.checkerframework.framework.util.AnnotatedTypes;
 import org.checkerframework.javacutil.AnnotationUtils;
@@ -26,17 +30,54 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
 
     @Override
     public AnnotatedTypeMirror visitVariable(VariableTree node, AnnotatedTypeFactory f) {
+        Element elt = TreeUtils.elementFromDeclaration(node);
+
         // Create the ATM and add non-primary annotations
-        // (node.getType() does not include primary annotations, those are in
-        // node.getModifier()
+        // (node.getType() does not include the annotation before the type, so those
+        // are added to the type below.
         AnnotatedTypeMirror result = TypeFromTree.fromTypeTree(f, node.getType());
 
-        // Add primary annotations
+        // Handle the annotation in node.getModifiers().
+        List<AnnotationMirror> modifierAnnos;
         List<? extends AnnotationTree> annoTrees = node.getModifiers().getAnnotations();
         if (annoTrees != null && !annoTrees.isEmpty()) {
-            List<AnnotationMirror> annos = TreeUtils.annotationsFromTypeAnnotationTrees(annoTrees);
+            modifierAnnos = TreeUtils.annotationsFromTypeAnnotationTrees(annoTrees);
+        } else {
+            modifierAnnos = new ArrayList<>();
+        }
+
+        if (result.getKind() == TypeKind.DECLARED
+                &&
+                // Annotations on enum constants are not in the TypeMirror and always apply to the
+                // inner most type, so handle them in the else block.
+                elt.getKind() != ElementKind.ENUM_CONSTANT) {
+
+            // Decode the annotations from the type mirror because the annotations are already in
+            // the correct place for enclosing types. For example, @Tainted Outer.Inner and @Tainted
+            // Inner. For both types @Tainted is store in node.getModifier(), but they apply to
+            // different types.
+            AnnotatedDeclaredType annotatedDeclaredType = (AnnotatedDeclaredType) result;
+            // The underlying type of result does not have all annotations, but the TypeMirror of
+            // node.getType() does.
+            DeclaredType declaredType = (DeclaredType) TreeUtils.typeOf(node.getType());
+            AnnotatedTypes.applyAnnotationsFromDeclaredType(annotatedDeclaredType, declaredType);
+
+            // Handle declaration annotations
+            for (AnnotationMirror anno : modifierAnnos) {
+                if (AnnotationUtils.isDeclarationAnnotation(anno)) {
+                    // This does not treat Checker Framework compatqual annotations differently,
+                    // becuase
+                    // it's not clear whether the annotation should apply to the outer most
+                    // enclosing type
+                    // or the inner most.
+                    result.addAnnotation(anno);
+                } // if anno is not a declaration annotations, it should have been applied in the
+                // call to applyAnnotationsFromDeclaredType above.
+            }
+        } else {
+            // Add  the primary annotation from the node.getModifiers();
             AnnotatedTypeMirror innerType = AnnotatedTypes.innerMostType(result);
-            for (AnnotationMirror anno : annos) {
+            for (AnnotationMirror anno : modifierAnnos) {
                 // The code here is similar to
                 // org.checkerframework.framework.util.element.ElementAnnotationUtil.addDeclarationAnnotationsFromElement.
                 if (AnnotationUtils.isDeclarationAnnotation(anno)
@@ -52,7 +93,6 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
             }
         }
 
-        Element elt = TreeUtils.elementFromDeclaration(node);
         AnnotatedTypeMirror lambdaParamType = inferLambdaParamAnnotations(f, result, elt);
         if (lambdaParamType != null) {
             return lambdaParamType;

--- a/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
@@ -66,10 +66,8 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
             for (AnnotationMirror anno : modifierAnnos) {
                 if (AnnotationUtils.isDeclarationAnnotation(anno)) {
                     // This does not treat Checker Framework compatqual annotations differently,
-                    // becuase
-                    // it's not clear whether the annotation should apply to the outer most
-                    // enclosing type
-                    // or the inner most.
+                    // because it's not clear whether the annotation should apply to the outermost
+                    // enclosing type or the innermost.
                     result.addAnnotation(anno);
                 } // if anno is not a declaration annotations, it should have been applied in the
                 // call to applyAnnotationsFromDeclaredType above.

--- a/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
@@ -34,10 +34,10 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
 
         // Create the ATM and add non-primary annotations
         // (node.getType() does not include the annotation before the type, so those
-        // are added to the type below.
+        // are added to the type below).
         AnnotatedTypeMirror result = TypeFromTree.fromTypeTree(f, node.getType());
 
-        // Handle the annotation in node.getModifiers().
+        // Handle any annotations in node.getModifiers().
         List<AnnotationMirror> modifierAnnos;
         List<? extends AnnotationTree> annoTrees = node.getModifiers().getAnnotations();
         if (annoTrees != null && !annoTrees.isEmpty()) {
@@ -49,12 +49,12 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
         if (result.getKind() == TypeKind.DECLARED
                 &&
                 // Annotations on enum constants are not in the TypeMirror and always apply to the
-                // inner most type, so handle them in the else block.
+                // innermost type, so handle them in the else block.
                 elt.getKind() != ElementKind.ENUM_CONSTANT) {
 
             // Decode the annotations from the type mirror because the annotations are already in
             // the correct place for enclosing types. For example, @Tainted Outer.Inner and @Tainted
-            // Inner. For both types @Tainted is store in node.getModifier(), but they apply to
+            // Inner. For both types @Tainted is stored in node.getModifiers(), but they apply to
             // different types.
             AnnotatedDeclaredType annotatedDeclaredType = (AnnotatedDeclaredType) result;
             // The underlying type of result does not have all annotations, but the TypeMirror of
@@ -69,11 +69,11 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
                     // because it's not clear whether the annotation should apply to the outermost
                     // enclosing type or the innermost.
                     result.addAnnotation(anno);
-                } // if anno is not a declaration annotations, it should have been applied in the
+                } // If anno is not a declaration annotation, it should have been applied in the
                 // call to applyAnnotationsFromDeclaredType above.
             }
         } else {
-            // Add  the primary annotation from the node.getModifiers();
+            // Add the primary annotation from the node.getModifiers();
             AnnotatedTypeMirror innerType = AnnotatedTypes.innerMostType(result);
             for (AnnotationMirror anno : modifierAnnos) {
                 // The code here is similar to

--- a/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
+++ b/framework/src/main/java/org/checkerframework/framework/type/TypeFromMemberVisitor.java
@@ -29,17 +29,17 @@ import org.checkerframework.javacutil.TypesUtils;
 class TypeFromMemberVisitor extends TypeFromTreeVisitor {
 
     @Override
-    public AnnotatedTypeMirror visitVariable(VariableTree node, AnnotatedTypeFactory f) {
-        Element elt = TreeUtils.elementFromDeclaration(node);
+    public AnnotatedTypeMirror visitVariable(VariableTree variableTree, AnnotatedTypeFactory f) {
+        Element elt = TreeUtils.elementFromDeclaration(variableTree);
 
         // Create the ATM and add non-primary annotations
-        // (node.getType() does not include the annotation before the type, so those
+        // (variableTree.getType() does not include the annotation before the type, so those
         // are added to the type below).
-        AnnotatedTypeMirror result = TypeFromTree.fromTypeTree(f, node.getType());
+        AnnotatedTypeMirror result = TypeFromTree.fromTypeTree(f, variableTree.getType());
 
-        // Handle any annotations in node.getModifiers().
+        // Handle any annotations in variableTree.getModifiers().
         List<AnnotationMirror> modifierAnnos;
-        List<? extends AnnotationTree> annoTrees = node.getModifiers().getAnnotations();
+        List<? extends AnnotationTree> annoTrees = variableTree.getModifiers().getAnnotations();
         if (annoTrees != null && !annoTrees.isEmpty()) {
             modifierAnnos = TreeUtils.annotationsFromTypeAnnotationTrees(annoTrees);
         } else {
@@ -53,13 +53,16 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
                 elt.getKind() != ElementKind.ENUM_CONSTANT) {
 
             // Decode the annotations from the type mirror because the annotations are already in
-            // the correct place for enclosing types. For example, @Tainted Outer.Inner and @Tainted
-            // Inner. For both types @Tainted is stored in node.getModifiers(), but they apply to
-            // different types.
+            // the correct place for enclosing types.  The annotations in
+            // variableTree.getModifiers()
+            // might apply to the enclosing type or the type itself. For example, @Tainted
+            // Outer.Inner y and @Tainted
+            // Inner x.  @Tainted is stored in variableTree.getModifiers() of the variable tree
+            // corresponding to both x and y, but @Tainted applies to different types.
             AnnotatedDeclaredType annotatedDeclaredType = (AnnotatedDeclaredType) result;
             // The underlying type of result does not have all annotations, but the TypeMirror of
-            // node.getType() does.
-            DeclaredType declaredType = (DeclaredType) TreeUtils.typeOf(node.getType());
+            // variableTree.getType() does.
+            DeclaredType declaredType = (DeclaredType) TreeUtils.typeOf(variableTree.getType());
             AnnotatedTypes.applyAnnotationsFromDeclaredType(annotatedDeclaredType, declaredType);
 
             // Handle declaration annotations
@@ -69,11 +72,12 @@ class TypeFromMemberVisitor extends TypeFromTreeVisitor {
                     // because it's not clear whether the annotation should apply to the outermost
                     // enclosing type or the innermost.
                     result.addAnnotation(anno);
-                } // If anno is not a declaration annotation, it should have been applied in the
-                // call to applyAnnotationsFromDeclaredType above.
+                }
+                // If anno is not a declaration annotation, it should have been applied in the call
+                // to applyAnnotationsFromDeclaredType above.
             }
         } else {
-            // Add the primary annotation from the node.getModifiers();
+            // Add the primary annotation from the variableTree.getModifiers();
             AnnotatedTypeMirror innerType = AnnotatedTypes.innerMostType(result);
             for (AnnotationMirror anno : modifierAnnos) {
                 // The code here is similar to

--- a/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
+++ b/framework/src/main/java/org/checkerframework/framework/util/AnnotatedTypes.java
@@ -27,6 +27,7 @@ import javax.lang.model.element.Element;
 import javax.lang.model.element.ExecutableElement;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.element.TypeParameterElement;
+import javax.lang.model.type.DeclaredType;
 import javax.lang.model.type.TypeKind;
 import javax.lang.model.type.TypeMirror;
 import javax.lang.model.type.TypeVariable;
@@ -1442,6 +1443,27 @@ public class AnnotatedTypes {
                     break;
                 }
             }
+        }
+    }
+
+    /**
+     * Add all the annotations in {@code declaredType} to {@code annotatedDeclaredType}.
+     *
+     * <p>(The {@code TypeMirror} returned by {@code annotatedDeclaredType#getUnderlyingType} may
+     * have not have all the annotations on the type, so allow the user to specify a different one.)
+     *
+     * @param annotatedDeclaredType annotated type to which annotations are added
+     * @param declaredType TypeMirror that may have annotations
+     */
+    public static void applyAnnotationsFromDeclaredType(
+            AnnotatedDeclaredType annotatedDeclaredType, DeclaredType declaredType) {
+        TypeMirror underlyingTypeMirror = declaredType;
+        while (annotatedDeclaredType != null) {
+            List<? extends AnnotationMirror> annosOnTypeMirror =
+                    underlyingTypeMirror.getAnnotationMirrors();
+            annotatedDeclaredType.addAnnotations(annosOnTypeMirror);
+            annotatedDeclaredType = annotatedDeclaredType.getEnclosingType();
+            underlyingTypeMirror = ((DeclaredType) underlyingTypeMirror).getEnclosingType();
         }
     }
 }

